File: server-node/src/config/firebase.js
import admin from 'firebase-admin';

export function initFirebase() {
  if (admin.apps.length) return admin.app();

  // Prefer JSON in env; fallback to GOOGLE_APPLICATION_CREDENTIALS file path
  const json = process.env.FIREBASE_SERVICE_ACCOUNT_JSON;
  if (json) {
    const credObj = JSON.parse(json);
    admin.initializeApp({
      credential: admin.credential.cert(credObj),
    });
  } else if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
    admin.initializeApp({
      credential: admin.credential.applicationDefault(),
    });
  } else {
    throw new Error('Firebase Admin credentials not provided. Set FIREBASE_SERVICE_ACCOUNT_JSON or GOOGLE_APPLICATION_CREDENTIALS');
  }
  return admin.app();
}

export function getFirebase() {
  if (!admin.apps.length) throw new Error('Firebase not initialized');
  return admin;
}

File: server-py/app/firebase.py
import json
import firebase_admin
from firebase_admin import credentials, auth
from .config import FIREBASE_SERVICE_ACCOUNT_JSON
import os

def init_firebase():
    if firebase_admin._apps:
        return firebase_admin.get_app()

    if FIREBASE_SERVICE_ACCOUNT_JSON:
        cred_obj = json.loads(FIREBASE_SERVICE_ACCOUNT_JSON)
        cred = credentials.Certificate(cred_obj)
    elif os.getenv('GOOGLE_APPLICATION_CREDENTIALS'):
        cred = credentials.ApplicationDefault()
    else:
        raise RuntimeError('Firebase Admin credentials not provided')

    return firebase_admin.initialize_app(cred)


def verify_id_token(id_token: str):
    init_firebase()
    return auth.verify_id_token(id_token)
